### ISSUE TITLE
Incorrect `unnecessary_cast` on a raw pointer with an unknown pointee type

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-clippy#11634]
### Summary

The lint should not warn in this case

### Lint Name

unnecessary_cast

### Reproducer

I tried this code:

```rust
fn src<A>() -> *mut *mut A {
    unimplemented!()
}

unsafe fn error_repo<A>() -> bool {
    let val = src();
    (*val as *mut A).is_null()
}
```

I saw this happen:

```
warning: casting raw pointers to the same type and constness is unnecessary (`*mut A` -> `*mut A`)
   --> src/functions.rs:622:5
    |
622 |     (*val as *mut A).is_null()
    |     ^^^^^^^^^^^^^^^^ help: try: `*val`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
    = note: `#[warn(clippy::unnecessary_cast)]` on by default
```

I expected not to see this error, because removing `as *mut A` fails to compile with this error:

```
error[E0699]: cannot call a method on a raw pointer with an unknown pointee type
   --> src/functions.rs:622:12
    |
622 |     (*val).is_null()
    |            ^^^^^^^
```

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```
